pkg/storage/engine/mvcc.go

```diff
@@ -2735,6 +2746,66 @@ func mvccResolveWriteIntent(
 	return true, nil
 }
 
+// mvccRewriteIntentHistory rolls back the write from the intent, ignoring
+// all values from the history that have an ignored seqnum.
+func mvccRewriteIntentHistory(
+	ctx context.Context,
+	engine ReadWriter,
+	txn *enginepb.TxnMeta,
+	meta *enginepb.MVCCMetadata,
+	latestKey MVCCKey,
+) (clear bool, err error) {
+	if !seqNumIsIgnored(txn, meta.Txn.Sequence) {
+		// The latest write was not ignored. Nothing to do here.  We'll
+		// proceed with the intent as usual.
+		return false, nil
+	}
+	// Find the latest historical write before that that was not
+	// ignored.
+	var i int
+	for i = len(meta.IntentHistory) - 1; i >= 0; i-- {
+		e := &meta.IntentHistory[i]
+		if !seqNumIsIgnored(txn, e.Sequence) {
+			break
+		}
+	}
+
+	// If i < 0, we don't have an intent any more: everything
+	// has been rolled back.
+	if i < 0 {
+		err := engine.Clear(latestKey)
+		return true, err
+	}
+
+	// Otherwise, we place back the write at that history entry
+	// back into the intent.
+	meta.Deleted = len(meta.IntentHistory[i].Value) == 0
```

Comment:
This line in particular makes me a little bit afraid.
The intent history does not distinguish between an empty value array, and a Delete intent. Does this matter?
@nvanbenschoten

Reply:
No, that doesn't matter. An empty value is a delete.